File: core/src/main/java/com/personal/image_toolkit/ImageFormatConverter.java
package com.personal.image_toolkit.tools;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.awt.Color;
import java.awt.Graphics2D;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * A tool for converting image formats for single files and batches.
 * Relies on FSETool for path management and directory creation.
 *
 * Re-implementation of FormatConverter.py
 */
public final class ImageFormatConverter {

    // This was imported in the Python version. We'll define it here.
    private static final Set<String> SUPPORTED_IMG_FORMATS = 
            Set.of("png", "jpg", "jpeg", "bmp", "gif");

    private ImageFormatConverter() {}

    /**
     * Core logic for image format conversion.
     * @return The converted image, or null on failure.
     */
    private static BufferedImage convertImgCore(String imagePath, String outputPath, String format, boolean delete) {
        File inputFile = new File(imagePath);
        File outputFile = new File(outputPath);
        String outputFormat = format.toLowerCase();

        // Check input extension
        String inputExt = getFileExtension(imagePath);
        if (!SUPPORTED_IMG_FORMATS.contains(inputExt)) {
            System.err.println("Invalid input file extension: " + inputExt);
            return null;
        }

        // Check output format
        if (!SUPPORTED_IMG_FORMATS.contains(outputFormat)) {
            System.err.println("Unsupported output format: " + outputFormat);
            return null;
        }

        try {
            BufferedImage img = ImageIO.read(inputFile);
            if (img == null) {
                throw new IOException("Could not read image file.");
            }

            // Handle JPEG conversion: ensure no transparency
            if ("jpg".equals(outputFormat) || "jpeg".equals(outputFormat)) {
                // Create a new image with a white background
                BufferedImage newImg = new BufferedImage(
                        img.getWidth(), img.getHeight(), BufferedImage.TYPE_INT_RGB
                );
                Graphics2D g = newImg.createGraphics();
                g.setColor(Color.WHITE); // Set background to white
                g.fillRect(0, 0, newImg.getWidth(), newImg.getHeight());
                g.drawImage(img, 0, 0, null);
                g.dispose();
                img = newImg;
            }
            
            // Write the new image
            ImageIO.write(img, outputFormat, outputFile);
            
            if (delete) {
                inputFile.delete();
            }
            System.out.println("Converted '" + inputFile.getName() + "' to '" + outputFile.getName() + "'.");
            return img;

        } catch (Exception e) {
            System.err.println("Warning: failed to convert file " + imagePath + ". Reason: " + e.getMessage());
            return null;
        }
    }

    /**
     * Converts a single image file to a specified format.
     */
    public static BufferedImage convertImgFormat(String imagePath, String outputName, String format, boolean delete) throws IOException {
        Path resolvedImagePath = FSETool.resolvePath(imagePath);
        String inputDir = resolvedImagePath.getParent().toString();
        String filenameOnly = getFileNameWithoutExtension(resolvedImagePath.getFileName().toString());

        String resolvedOutputPath;
        if (outputName == null || outputName.isEmpty()) {
            // Save in the same directory as the input image
            resolvedOutputPath = new File(inputDir, filenameOnly + "." + format).getAbsolutePath();
        } else {
            // outputName might be a full path or just a name.
            // FSETool.resolvePath handles both.
            Path outPath = FSETool.resolvePath(outputName);
            // If outputName was just a name, it's resolved relative to CWD.
            // Let's check if it has a parent. If not, place it in input dir.
            if (outPath.getParent() == null || outPath.getParent().toString().equals(System.getProperty("user.dir"))) {
                 resolvedOutputPath = new File(inputDir, outputName + "." + format).getAbsolutePath();
            } else {
                 resolvedOutputPath = outPath.toString() + "." + format;
            }
        }
        
        // Replaced decorator: Explicitly create parent directory
        FSETool.createDirectoryForFile(resolvedOutputPath);

        if (Files.isRegularFile(Paths.get(resolvedOutputPath))) {
            return null; // File already exists
        }

        return convertImgCore(resolvedImagePath.toString(), resolvedOutputPath, format, delete);
    }

    /**
     * Converts all images in a directory matching input_formats to the output_format.
     */
    public static List<BufferedImage> batchConvertImgFormat(
            String inputDir, List<String> inputsFormats, String outputDir, 
            String outputFormat, boolean delete) throws IOException {
        
        String resolvedOutputDir = (outputDir == null || outputDir.isEmpty()) 
                ? FSETool.resolvePath(inputDir).toString() 
                : FSETool.resolvePath(outputDir).toString();
        
        // Replaced decorator: Explicitly create output directory
        FSETool.createDirectory(resolvedOutputDir);

        String outFormat = outputFormat.toLowerCase();
        
        List<BufferedImage> newImages = new ArrayList<>();
        for (String inputFormat : inputsFormats) {
            if (inputFormat.toLowerCase().equals(outFormat)) {
                continue; // Skip if in-format == out-format
            }
            
            List<String> filesToConvert = FSETool.getFilesByExtension(inputDir, inputFormat, false);
            
            for (String inputFile : filesToConvert) {
                String filename = getFileNameWithoutExtension(new File(inputFile).getName());
                String outputPath = new File(resolvedOutputDir, filename + "." + outFormat).getAbsolutePath();
                
                if (!Files.isRegularFile(Paths.get(outputPath))) {
                    BufferedImage img = convertImgCore(inputFile, outputPath, outFormat, delete);
                    if (img != null) {
                        newImages.add(img);
                    }
                }
            }
        }
        
        System.out.println("\nBatch conversion complete! Converted " + newImages.size() + " images.");
        return newImages;
    }

    // Helper method
    private static String getFileExtension(String filename) {
        int dotIndex = filename.lastIndexOf('.');
        return (dotIndex == -1) ? "" : filename.substring(dotIndex + 1).toLowerCase();
    }

    // Helper method
    private static String getFileNameWithoutExtension(String filename) {
        int dotIndex = filename.lastIndexOf('.');
        return (dotIndex == -1) ? filename : filename.substring(0, dotIndex);
    }
}

File: backend/src/core/ImageFormatConverter.cpp
#include "ImageFormatConverter.hpp"
#include "FileSystemTool.hpp"
#include <iostream>
#include <filesystem>
#include <algorithm>

namespace fs = std::filesystem;

// Helper to check if format is jpg
bool isJpeg(const std::string& fmt) {
    std::string f = fmt;
    std::transform(f.begin(), f.end(), f.begin(), ::tolower);
    return f == "jpg" || f == "jpeg";
}

cv::Mat ImageFormatConverter::removeAlphaChannel(const cv::Mat& src) {
    if (src.channels() != 4) return src;

    // Create white background
    cv::Mat bg(src.size(), CV_8UC3, cv::Scalar(255, 255, 255));
    
    // Split source
    std::vector<cv::Mat> channels;
    cv::split(src, channels);
    
    cv::Mat alpha = channels[3];
    cv::Mat rgb;
    cv::merge(std::vector<cv::Mat>{channels[0], channels[1], channels[2]}, rgb);
    
    // Blend: dst = src * alpha + bg * (1 - alpha)
    alpha.convertTo(alpha, CV_32F, 1.0/255.0);
    rgb.convertTo(rgb, CV_32F);
    bg.convertTo(bg, CV_32F);
    
    cv::Mat dst = cv::Mat::zeros(src.size(), CV_32FC3);
    
    for (int i = 0; i < 3; ++i) {
        cv::Mat src_c, bg_c;
        cv::extractChannel(rgb, src_c, i);
        cv::extractChannel(bg, bg_c, i);
        
        cv::Mat res = src_c.mul(alpha) + bg_c.mul(1.0 - alpha);
        cv::insertChannel(res, dst, i);
    }
    
    dst.convertTo(dst, CV_8UC3);
    return dst;
}

bool ImageFormatConverter::convertImgCore(const std::string& image_path, 
                                          const std::string& output_path, 
                                          const std::string& format, 
                                          bool delete_original) {
    try {
        cv::Mat img = cv::imread(image_path, cv::IMREAD_UNCHANGED);
        if (img.empty()) {
            std::cerr << "Warning: Failed to load " << image_path << std::endl;
            return false;
        }

        // Handle transparency for JPEGs
        if (isJpeg(format) && img.channels() == 4) {
            img = removeAlphaChannel(img);
        }

        // Ensure output directory exists (mimics prefix_create_directory)
        FSETool::createDirectoryForFile(output_path);

        if (cv::imwrite(output_path, img)) {
            std::cout << "Converted '" << fs::path(image_path).filename().string() 
                      << "' to '" << fs::path(output_path).filename().string() << "'." << std::endl;
            if (delete_original) {
                FileDeleter::deletePath(image_path);
            }
            return true;
        }
        return false;
    } catch (const std::exception& e) {
        std::cerr << "Conversion Error: " << e.what() << std::endl;
        return false;
    }
}

bool ImageFormatConverter::convertSingleImage(const std::string& image_path, 
                                              const std::string& output_name, 
                                              const std::string& format, 
                                              bool delete_original) {
    std::string abs_path = FSETool::toAbsolutePath(image_path);
    if (!fs::exists(abs_path)) return false;

    fs::path p(abs_path);
    std::string out_path;

    if (output_name.empty()) {
        out_path = (p.parent_path() / (p.stem().string() + "." + format)).string();
    } else {
        // If output_name is just a name, append format. If it has a path, use it.
        fs::path out_p(output_name);
        if (!out_p.has_extension()) out_p.replace_extension("." + format);
        out_path = out_p.string();
    }

    if (fs::exists(out_path)) return false;
    return convertImgCore(abs_path, out_path, format, delete_original);
}

std::vector<std::string> ImageFormatConverter::convertBatch(const std::string& input_dir, 
                                                            const std::vector<std::string>& input_formats, 
                                                            const std::string& output_dir, 
                                                            const std::string& output_format, 
                                                            bool delete_original) {
    std::vector<std::string> converted_files;
    std::string in_dir_abs = FSETool::toAbsolutePath(input_dir);
    std::string out_dir_abs = output_dir.empty() ? in_dir_abs : FSETool::toAbsolutePath(output_dir);

    for (const auto& fmt : input_formats) {
        // Skip converting same format
        if (fmt == output_format) continue; 
        if (isJpeg(fmt) && isJpeg(output_format)) continue;

        auto files = FSETool::getFilesByExtension(in_dir_abs, fmt);
        for (const auto& file : files) {
            fs::path p(file);
            fs::path out_p = fs::path(out_dir_abs) / (p.stem().string() + "." + output_format);
            
            if (!fs::exists(out_p)) {
                if (convertImgCore(file, out_p.string(), output_format, delete_original)) {
                    converted_files.push_back(out_p.string());
                }
            }
        }
    }
    std::cout << "\nBatch conversion complete! Converted " << converted_files.size() << " images." << std::endl;
    return converted_files;
}